File: Image_Processing_Algorithms/NLM/opencv_implementation/nlm_opencv.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/photo.hpp>
#include <iostream>

class NLMDenoising {
private:
    int templateWindowSize;
    int searchWindowSize;
    float h;
    bool useMultiThreading;

public:
    NLMDenoising(int templateWindow = 7, int searchWindow = 21, float hParam = 10.0f, 
                 bool multiThread = true) 
        : templateWindowSize(templateWindow), 
          searchWindowSize(searchWindow),
          h(hParam),
          useMultiThreading(multiThread) {}

    cv::Mat denoise(const cv::Mat& input) {
        CV_Assert(!input.empty() && input.channels() <= 4);
        
        cv::Mat output;
        if (input.channels() == 1) {
            // Grayscale image processing
            cv::fastNlMeansDenoising(input, output, h, templateWindowSize, searchWindowSize);
        } else {
            // Color image processing
            cv::fastNlMeansDenoisingColored(input, output, h, h, templateWindowSize, searchWindowSize);
        }
        
        return output;
    }

    cv::Mat denoiseCustom(const cv::Mat& input) {
        CV_Assert(!input.empty() && input.channels() <= 4);
        
        cv::Mat output = input.clone();
        cv::Mat inputFloat;
        input.convertTo(inputFloat, CV_32F, 1.0/255.0);

        // Parameters
        const int halfTemplate = templateWindowSize / 2;
        const int halfSearch = searchWindowSize / 2;
        const float hSquared = h * h;

        // Parallel processing if enabled
        #pragma omp parallel for if(useMultiThreading) collapse(2)
        for (int i = halfTemplate; i < input.rows - halfTemplate; i++) {
            for (int j = halfTemplate; j < input.cols - halfTemplate; j++) {
                cv::Mat patch = inputFloat(
                    cv::Range(i - halfTemplate, i + halfTemplate + 1),
                    cv::Range(j - halfTemplate, j + halfTemplate + 1)
                );

                float weightSum = 0;
                cv::Vec3f pixelSum = cv::Vec3f(0, 0, 0);

                // Search window
                for (int si = -halfSearch; si <= halfSearch; si++) {
                    for (int sj = -halfSearch; sj <= halfSearch; sj++) {
                        int ni = i + si;
                        int nj = j + sj;

                        if (ni < halfTemplate || ni >= input.rows - halfTemplate ||
                            nj < halfTemplate || nj >= input.cols - halfTemplate) {
                            continue;
                        }

                        cv::Mat neighborPatch = inputFloat(
                            cv::Range(ni - halfTemplate, ni + halfTemplate + 1),
                            cv::Range(nj - halfTemplate, nj + halfTemplate + 1)
                        );

                        // Compute patch distance
                        cv::Mat diff = patch - neighborPatch;
                        float distance = cv::norm(diff, cv::NORM_L2SQR) / (templateWindowSize * templateWindowSize);
                        float weight = std::exp(-distance / hSquared);

                        weightSum += weight;
                        pixelSum += weight * inputFloat.at<cv::Vec3f>(ni, nj);
                    }
                }

                output.at<cv::Vec3b>(i, j) = cv::Vec3b(
                    cv::saturate_cast<uchar>(pixelSum[0] * 255.0f / weightSum),
                    cv::saturate_cast<uchar>(pixelSum[1] * 255.0f / weightSum),
                    cv::saturate_cast<uchar>(pixelSum[2] * 255.0f / weightSum)
                );
            }
        }

        return output;
    }

    // Utility function to compare OpenCV's implementation with custom implementation
    void compareImplementations(const cv::Mat& input) {
        cv::Mat opencvResult = denoise(input);
        cv::Mat customResult = denoiseCustom(input);

        cv::Mat diff;
        cv::absdiff(opencvResult, customResult, diff);

        cv::Scalar meanDiff = cv::mean(diff);
        std::cout << "Average difference between implementations: "
                  << (meanDiff[0] + meanDiff[1] + meanDiff[2])/3.0 << std::endl;

        // Display results
        cv::imshow("Original", input);
        cv::imshow("OpenCV NLM", opencvResult);
        cv::imshow("Custom NLM", customResult);
        cv::imshow("Difference", diff);
        cv::waitKey(0);
    }
};

// Example usage
int main(int argc, char** argv) {
    if (argc != 2) {
        std::cout << "Usage: " << argv[0] << " <image_path>" << std::endl;
        return -1;
    }

    // Load image
    cv::Mat input = cv::imread(argv[1]);
    if (input.empty()) {
        std::cout << "Error: Could not read image." << std::endl;
        return -1;
    }

    // Create NLM denoising object
    NLMDenoising nlm(7, 21, 10.0f, true);

    // Process and compare implementations
    nlm.compareImplementations(input);

    return 0;
}

File: Image_Processing_Algorithms/NLM/opencv_implementation/nlm_opencv.py
import cv2
import numpy as np

class NLMDenoising:
    def __init__(self, template_window=7, search_window=21, h_param=10.0, multi_thread=True):
        self.template_window_size = template_window
        self.search_window_size = search_window
        self.h = h_param
        self.use_multi_threading = multi_thread

    def denoise(self, input_image):
        assert input_image is not None and input_image.shape[2] <= 4
        if len(input_image.shape) == 2:
            # Grayscale image processing
            return cv2.fastNlMeansDenoising(input_image, None, self.h, self.template_window_size, self.search_window_size)
        else:
            # Color image processing
            return cv2.fastNlMeansDenoisingColored(input_image, None, self.h, self.h, self.template_window_size, self.search_window_size)

    def denoise_custom(self, input_image):
        assert input_image is not None and input_image.shape[2] <= 4
        input_float = input_image.astype(np.float32) / 255.0
        output = input_float.copy()

        half_template = self.template_window_size // 2
        half_search = self.search_window_size // 2
        h_squared = self.h * self.h

        for i in range(half_template, input_float.shape[0] - half_template):
            for j in range(half_template, input_float.shape[1] - half_template):
                patch = input_float[i - half_template:i + half_template + 1, j - half_template:j + half_template + 1]
                weight_sum = 0
                pixel_sum = np.zeros(3)

                for si in range(-half_search, half_search + 1):
                    for sj in range(-half_search, half_search + 1):
                        ni, nj = i + si, j + sj
                        if ni < half_template or ni >= input_float.shape[0] - half_template or nj < half_template or nj >= input_float.shape[1] - half_template:
                            continue

                        neighbor_patch = input_float[ni - half_template:ni + half_template + 1, nj - half_template:nj + half_template + 1]
                        distance = np.sum((patch - neighbor_patch) ** 2) / (self.template_window_size ** 2)
                        weight = np.exp(-distance / h_squared)

                        weight_sum += weight
                        pixel_sum += weight * input_float[ni, nj]

                output[i, j] = np.clip(pixel_sum * 255.0 / weight_sum, 0, 255)

        return output.astype(np.uint8)

    def compare_implementations(self, input_image):
        opencv_result = self.denoise(input_image)
        custom_result = self.denoise_custom(input_image)

        mean_diff = np.mean(np.abs(opencv_result - custom_result))
        print(f"Average difference between implementations: {mean_diff}")

        cv2.imshow("Original", input_image)
        cv2.imshow("OpenCV NLM", opencv_result)
        cv2.imshow("Custom NLM", custom_result)
        cv2.waitKey(0)

# Example usage
if __name__ == '__main__':
    # Load an image
    image = cv2.imread('/home/jatayu/MyFiles/Samsung_Prism_NEON/images/input_images/BM3D/Basic3.jpg')  # Replace with actual image path
    nlm = NLMDenoising()
    nlm.compare_implementations(image)
